go/tools/builders/nogo_change.go

```diff
@@ -0,0 +1,326 @@
+// Change represents a set of edits to be applied to a set of files.
+type Change struct {
+	AnalyzerToFileToEdits map[string]map[string][]Edit `json:"analyzer_file_to_edits"`
+}
+
+// NewChange creates a new Change object.
+func NewChange() *Change {
+	return &Change{
+		AnalyzerToFileToEdits: make(map[string]map[string][]Edit),
+	}
+}
+
+// NewChangeFromDiagnostics builds a Change from a set of diagnostics.
+// Unlike Diagnostic, Change is independent of the FileSet given it uses perf-file offsets instead of token.Pos.
+// This allows Change to be used in contexts where the FileSet is not available, e.g., it remains applicable after it is saved to disk and loaded back.
+// See https://github.com/golang/tools/blob/master/go/analysis/diagnostic.go for details.
+func NewChangeFromDiagnostics(entries []DiagnosticEntry, fileSet *token.FileSet) (*Change, error) {
+	c := NewChange()
+
+	cwd, err := os.Getwd() // workspace root
+	if err != nil {
+		return c, fmt.Errorf("Error getting current working directory: (%v)", err)
+	}
+
+	var allErrors []error
+
+	for _, entry := range entries {
+		analyzer := entry.Analyzer.Name
+		for _, sf := range entry.Diagnostic.SuggestedFixes {
+			for _, edit := range sf.TextEdits {
+				start, end := edit.Pos, edit.End
+				if !end.IsValid() {
+					// In insertion, end could be token.NoPos
+					end = start
+				}
+
+				file := fileSet.File(edit.Pos)
+				if file == nil {
+					allErrors = append(allErrors, fmt.Errorf("invalid fix: missing file info for pos %v", edit.Pos))
+					continue
+				}
+				if start > end {
+					allErrors = append(allErrors, fmt.Errorf("invalid fix: pos %v > end %v", start, end))
+					continue
+				}
+				if eof := token.Pos(file.Base() + file.Size()); end > eof {
+					allErrors = append(allErrors, fmt.Errorf("invalid fix: end %v past end of file %v", end, eof))
+					continue
+				}
+
+				edit := Edit{Start: file.Offset(start), End: file.Offset(end), New: string(edit.NewText)}
+				fileRelativePath, err := filepath.Rel(cwd, file.Name())
+				if err != nil {
+					fileRelativePath = file.Name() // fallback logic
+				}
+				c.AddEdit(analyzer, fileRelativePath, edit)
+			}
+		}
+	}
+
+	if len(allErrors) > 0 {
+		return c, fmt.Errorf("errors: %v", allErrors)
+	}
+	return c, nil
+}
+
+// AddEdit adds an edit to the change.
+func (c *Change) AddEdit(analyzer string, file string, edit Edit) {
+	// Check if the analyzer exists in the map
+	if _, ok := c.AnalyzerToFileToEdits[analyzer]; !ok {
+		// Initialize the map for the analyzer if it doesn't exist
+		c.AnalyzerToFileToEdits[analyzer] = make(map[string][]Edit)
+	}
+
+	// Append the edit to the list of edits for the specific file under the analyzer
+	c.AnalyzerToFileToEdits[analyzer][file] = append(c.AnalyzerToFileToEdits[analyzer][file], edit)
+}
+
+// Flatten takes a Change and returns a map of FileToEdits, merging edits from all analyzers.
+func Flatten(change Change) map[string][]Edit {
```

Comment:
Can we produce one patch file per analyzer? That way, we don't need to merge edits to the same file from different analyzers. Like you said, some analyzers may produce bad edits. When users apply patches one by one, they can ignore those patched produced by bad analyzers.

Reply:
there will be a single file per target, as created at the .bzl files (the go files only fill in contents).
but still, in that single file, we can index patches with analyzer. 

asking users to pick which analyzer to patch may be too much for developers. Ideally we want them simply say "yes/no". Also the conflicts may not happen often, we may ask only in the case of conflicts.

will think further.

Reply:
done, the patching is moved out of rules_go, 

fyi, it has the support of letting users preview the diff and pick. At the end, it applies the selected edits. In case of overlapping, we skip the first overlapping one for max progress, the next build can fix that one too.

Reply:
> asking users to pick which analyzer to patch may be too much for developers.

I don't think so. The output from rules_go can be simply:

```
nogo: errors found by nogo during build-time code analysis:
src/example.com/foo/compute.go:28:2: self-assignment of x to x (assign)

To apply the fixes run:

$ patch -p1 < bazel-out/k8-fastbuild/bin/src/example.com/foo/go_default_library.nogo.assign.patch
```

When there are multiple analyzers failing, we can list multiple patch command for users to copy:

```
To apply the fixes run:

$ patch -p1 < bazel-out/k8-fastbuild/bin/src/example.com/foo/go_default_library.nogo.assign.patch
$ patch -p1 < bazel-out/k8-fastbuild/bin/src/example.com/foo/go_default_library.nogo.composite.patch
```

I agree with what @fmeum said here: https://github.com/bazel-contrib/rules_go/pull/4102#discussion_r1779996815

I don't want rules_go to depend on a custom patch tool, while we can use a standard Unix utility.

> Ideally we want them simply say "yes/no".

I don't think users' input matters there. When two fixes conflict, there is nothing the tool can do even the user reply "yes", right? The only thing we can do is to reject the patch and ask user to manually apply the fix, which is exactly what the standard `patch` command does.

Reply:
discussed offline. 

agreed that we should use standard tool like 'patch -p1'.

we cannot have multiple patch files, each for one analyzer, since the bazel creates a single file for each target. Also it is not easy to tell the analyzers at bazel resolving time. 

what we currently do is:
- for each file, we apply the edits (offset-based) and drop the analyzers that introduce overlapping edits with added edits. Analyzers are sorted for order. After this, we can compute the new content after fixing. Also, we can compute the patch from orig and new content. 
- we merge patches for all files into one, so that "patch -p1 < file" works.

users' inputs do not matter, for the analyzers they pick, they may still have overlapping edits that we need to resolve.